### PR TITLE
[MSE] screen should sleep if the video has stalled, not just paused

### PR DESCRIPTION
**Pull request by jyavenard**
#### eaa411f285640e50b506f871981eed87e6e91fa3
<pre>
[MSE] screen should sleep if the video has stalled, not just paused
<a href="https://bugs.webkit.org/show_bug.cgi?id=270620">https://bugs.webkit.org/show_bug.cgi?id=270620</a>
<a href="https://rdar.apple.com/124190917">rdar://124190917</a>

Reviewed by Youenn Fablet.

A video may be playing as far as the HTMLMediaElement is concerned, but has stalled.
Those videos should allow the screen to sleep.

The code relied on an undocumented requirement that MediaPlayer::paused() returned
true if the video has stalled, but this requirement causes breakage elsewhere (see
webkit.org/b/270614)

Checking on the value of MediaPlayerPrivateRemote::timeIsProgressing instead
of paused() made several tests fail. Each failure revealed a discrepency
between the GPU&apos;s MediaPlayer&apos;s time and the simulated currentTime in the WP.
To avoid those, we re-sync the time estimator in each messages sent by the
GPU process (currentTime, wall time, and if media player&apos;s time is progressing).

No change in observable behaviour at this stage.
Covered by existing tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::shouldDisableSleep const):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::timeIsProgressing const):
(WebCore::MediaPlayer::currentTimeMayProgress const): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::timeIsProgressing const):
(WebCore::MediaPlayerPrivateInterface::currentTimeMayProgress const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::timeIsProgressing const):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::currentTimeMayProgress const): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::timeIsProgressing const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::timeIsProgressing const):
(WebCore::MediaPlayerPrivateGStreamerMSE::currentTimeMayProgress const): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::timeIsProgressing const):
(WebCore::MockMediaPlayerMediaSource::currentTimeMayProgress const): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::~RemoteMediaPlayerProxy):
(WebKit::timeUpdateData const):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerSeeked):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerTimeChanged):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerRateChanged):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerPlaybackStateChanged):
(WebKit::RemoteMediaPlayerProxy::currentTimeChanged):
(WebKit::RemoteMediaPlayerProxy::performTaskAtTime): Code was taking into account potential delay
between the time the WP sent a message and the GPUP received it. However
the time given to run the task is in the GPUP&apos;s time referential already
and as such, shouldn&apos;t be adjusted. Additionally, if the time sent had
already been in the past, it would have caused the simulated time to also
go backward which was incorrect.
(WebKit::RemoteMediaPlayerProxy::mediaPlayerPausedOrStalled const): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setTime):
(WebKit::MediaPlayerPrivateRemote::timeIsProgressing const):
(WebKit::MediaPlayerPrivateRemote::seekToTarget):
(WebKit::MediaPlayerPrivateRemote::seeked):
(WebKit::MediaPlayerPrivateRemote::timeChanged):
(WebKit::MediaPlayerPrivateRemote::rateChanged):
(WebKit::MediaPlayerPrivateRemote::playbackStateChanged):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):
(WebKit::MediaPlayerPrivateRemote::performTaskAtTime):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaTimeUpdateData.serialization.in: Added.

Canonical link: <a href="https://commits.webkit.org/275831@main">https://commits.webkit.org/275831@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0c0f55e30ab6f722615bcd47ddf1fbfe7458ca60

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45617 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39118 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45299 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/25755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19442 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19051 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37019 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16546 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/42864 "Passed tests") | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16643 "7 flakes 5 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38078 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1049 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39174 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38409 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47147 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14693 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42326 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19445 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40978 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9577 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19624 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19075 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/MediaPlayerPrivate.h [hunk 1]
```diff
@@ -148,6 +148,7 @@ class MediaPlayerPrivateInterface {
     virtual void setPitchCorrectionAlgorithm(MediaPlayer::PitchCorrectionAlgorithm) { }
 
     virtual bool paused() const = 0;
+    virtual bool timeIsProgressing() const { return paused(); }
```
Comment (youennf) [inline Source/WebCore/platform/graphics/MediaPlayerPrivate.h]: That is strange, should it be `!paused()`?

File: Source/WebCore/html/HTMLMediaElement.cpp [hunk 2]
```diff
@@ -7737,7 +7737,7 @@ HTMLMediaElement::SleepType HTMLMediaElement::shouldDisableSleep() const
 #if !PLATFORM(COCOA) && !PLATFORM(GTK) && !PLATFORM(WPE)
     return SleepType::None;
 #endif
-    if (m_sentEndEvent || !m_player || m_player->paused() || loop())
+    if (m_sentEndEvent || !m_player || !m_player->timeIsProgressing() || loop())
```
Comment (youennf) [inline Source/WebCore/html/HTMLMediaElement.cpp:7740]: It seems we should be able to write a test that checks shouldDisableSleep via internals API.
Can we do that for the various players if this is not already the case?

File: Source/WebCore/platform/graphics/MediaPlayerPrivate.h (hunk 1 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/MediaPlayerPrivate.h]: it should be yes !

File: Source/WebCore/html/HTMLMediaElement.cpp (hunk 2 above)
Comment (jyavenard) [reply Source/WebCore/html/HTMLMediaElement.cpp:7740]: we have: already `media/media-source/media-source-stalled-holds-sleep-assertion.html` it's this test that started to fail after PR#25565

Comment (youennf) [reply Source/WebCore/html/HTMLMediaElement.cpp:7740]: OK, I see tests for MSE and AVFoundation. We might be missing tests for webm and media stream.
I filed https://bugs.webkit.org/show_bug.cgi?id=270625.

**Review by youennf — APPROVED** [review]
LGTM.
Could you check whether we have coverage for WebM?
I can check MediaStream if you prefer.

**Comment by jyavenard** [comment]
> LGTM. Could you check whether we have coverage for WebM? I can check MediaStream if you prefer.

WebM player as it is can't stall.

**Pull request merged**